quickwit-search/src/root.rs

```diff
@@ -110,67 +121,85 @@ pub struct ErrorRetries {
 /// - Is the node ok, but just the split failing?
 /// - Did all requests fail? (Should we retry in that case?)
 fn analyze_errors(search_result: &SearchResultsByAddr) -> Option<ErrorRetries> {
-    let contains_errors = search_result.values().any(|res| {
-        res.is_err()
-            || res
-                .as_ref()
-                .map(|ok_res| ok_res.failed_requests.is_empty())
-                .unwrap_or(false)
-    });
-    if !contains_errors {
+    // Here we collect the failed requests on the node. It does not yet include failed requests
+    // against that node
+    let mut retry_split_ids = search_result
+        .values()
+        .filter_map(|result| result.as_ref().ok())
+        .flat_map(|res| {
+            res.failed_splits
+                .iter()
+                .filter(move |failed_splits| failed_splits.retryable_error) // only retry splits marked as retryable
+                .map(move |failed_splits| failed_splits.split_id.to_string())
+        })
+        .collect_vec();
+
+    // Include failed requests against that node
+    let failed_splits = search_result
+        .values()
+        .filter_map(|result| result.as_ref().err())
+        .flat_map(|err| err.split_ids.iter().cloned());
+    retry_split_ids.extend(failed_splits);
+
+    let contains_retryable_error = retry_split_ids.is_empty();
+    if contains_retryable_error {
         return None;
     }
+
     let complete_failure_nodes: Vec<_> = search_result
         .iter()
-        .filter(|(_addr, result)| {
-            if let Ok(ok_res) = result {
-                ok_res.failed_requests.len() as u64 == ok_res.num_attempted_splits
-            } else {
-                true
-            }
-        })
+        .filter(|(_addr, result)| is_complete_failure(*result))
         .map(|(addr, _)| *addr)
         .collect();
     let partial_or_no_failure_nodes: Vec<_> = search_result
         .iter()
         .filter(|(_addr, result)| {
             result
                 .as_ref()
-                .map(|ok_res| ok_res.failed_requests.len() as u64 != ok_res.num_attempted_splits)
+                .map(|ok_res| ok_res.failed_splits.len() as u64 != ok_res.num_attempted_splits)
```

Comment:
isn't it `!is_complete_failure()`? If it is we should use it, or even use .partition(...).

Reply:
yes, that's much better